### ISSUE TITLE
Error while generating virtual frame product CPF_L1VFRA

### ISSUE THREAD

**Issue by lmar76** [stcorp/procsim#35]
I'm trying to simulate the processor generating the CPF_L1VFRA product. The scenario configuration is the following:

```json
{
    "mission": "biomass",
    "scenarios": [
        {
            "name": "Framer",
            "file_name": "framer",
            "processor_name": "Framer",
            "processor_version": "01.00",
            "task_name": "Framer",
            "task_version": "01.00",
            "log_level": "debug",
            "begin_end_position_from_toi": true,
            "toi_start_offset": 0,
            "toi_stop_offset": 0,
            "anx": [
                "2017-02-25T05:00:18.883013Z",
                "2017-02-25T06:38:29.864514Z",
                "2017-02-25T08:16:40.846015Z",
                "2017-02-25T09:54:51.827516Z",
                "2017-02-25T11:33:02.809016Z",
                "2017-02-25T13:11:13.790517Z"
            ],
            "outputs": [
                {
                    "type": "CPF_L1VFRA",
                    "output_path": "outputs/CPF_L1VFRA"
                }
            ]
        }
    ]
}
```

The `TOI` in the job order corresponds to the sensing start/stop of the input slice `S3_RAW__0S` product.

Running procsim I got the following messages:

```
(pnode)  % procsim -t framer -j JobOrder.xml framer-scenario.json
2022-05-25T08:35:11.132838+00:00 ESR01104747 Framer 01.00 Framer [0000032228]: [I] Procsim v1.7.0 processor stub simulator
2022-05-25T08:35:11.133052+00:00 ESR01104747 Framer 01.00 Framer [0000032228]: [I] Simulate scenario Framer
2022-05-25T08:35:11.133064+00:00 ESR01104747 Framer 01.00 Framer [0000032228]: [I] Read JobOrder JobOrder.xml
2022-05-25T08:35:11.133073+00:00 ESR01104747 Framer 01.00 Framer [0000032228]: [D] JobOrder validation against schema: OK
2022-05-25T08:35:11.133081+00:00 ESR01104747 Framer 01.00 Framer [0000032228]: [I] Read task Framer from the JobOrder
2022-05-25T08:35:11.133094+00:00 ESR01104747 Framer 01.00 Framer [0000032228]: [I] Input type S3_RAW__0S: BIO_S3_RAW__0S_20170225T094604_20170225T094703_I_G01_M02_C01_T000_F057_01_BOSSJH
2022-05-25T08:35:11.133104+00:00 ESR01104747 Framer 01.00 Framer [0000032228]: [I] Input type S3_RAW__0M: BIO_S3_RAW__0M_20170225T094604_20170225T095834_I_G01_M02_C01_T000_F____01_BOSSJH
2022-05-25T08:35:11.133112+00:00 ESR01104747 Framer 01.00 Framer [0000032228]: [I] Input type AUX_ORB___: BIO_AUX_ORB____20170225T094548_20170225T095834_01_BOSSJH
2022-05-25T08:35:11.157583+00:00 ESR01104747 Framer 01.00 Framer [0000032228]: [D] Overwrite begin_end_position_from_toi from False to True
2022-05-25T08:35:11.157623+00:00 ESR01104747 Framer 01.00 Framer [0000032228]: [D] Overwrite toi_start_offset from 0.0 to 0.0
2022-05-25T08:35:11.157637+00:00 ESR01104747 Framer 01.00 Framer [0000032228]: [D] Overwrite toi_stop_offset from 0.0 to 0.0
2022-05-25T08:35:11.157653+00:00 ESR01104747 Framer 01.00 Framer [0000032228]: [D] Overwrite output_path from outputs/CPF_L1VFRA to outputs/CPF_L1VFRA
2022-05-25T08:35:11.157726+00:00 ESR01104747 Framer 01.00 Framer [0000032228]: [D] Output directory is /opt/pnode/data/wdir-s57/outputs/CPF_L1VFRA
2022-05-25T08:35:11.157745+00:00 ESR01104747 Framer 01.00 Framer [0000032228]: [D] Use begin_position as input for validity start time
2022-05-25T08:35:11.157760+00:00 ESR01104747 Framer 01.00 Framer [0000032228]: [D] Use end_position as input for validity stop time
2022-05-25T08:35:11.157789+00:00 ESR01104747 Framer 01.00 Framer [0000032228]: [E] Cannot perform framing, slice nr. is not known
2022-05-25T08:35:11.157804+00:00 ESR01104747 Framer 01.00 Framer [0000032228]: [I] Terminate with code 1
```

Please find attached the working directory including the scenario configuration:
[wdir-s57.zip](https://github.com/stcorp/procsim/files/8769419/wdir-s57.zip)

**Comment by mivermeu** [comment]
Thank you for the extensive description and attached working directory. It really helps to resolve these issues.

We were able to reproduce the issue and solve it. Previously, a slice number was supposed to be supplied ('slice_frame_nr' in the scenario) so that the frame numbers could be determined from it. That was not changed with the implementation of these virtual frames.

You can now specify just the slice bounds and ANX times, as you have done in your attached scenario, and procsim will automatically determine the slice number.

Let us know at what point you would like us to make another release. I assume you would like to do more testing soon, so perhaps it is best to hold off until you have verified the new feature works as expected.

**Comment by lmar76** [comment]
Hello, I tried again. Now procsim generates 3 outputs in the time interval [2017-02-25T09:46:04,  2017-02-25T09:46:58]:

- BIO_OPER_CPF_L1VFRA_20170225T094604_20170225T094620_01_BOUPMV.EOF
- BIO_OPER_CPF_L1VFRA_20170225T094618_20170225T094639_01_BOUPMV.EOF
- BIO_OPER_CPF_L1VFRA_20170225T094637_20170225T094658_01_BOUPMV.EOF

But the input L0 slice (`S3_RAW__0S`) has sensing interval [2017-02-25T09:46:04.620, 2017-02-25T09:47:03.747]. So, I was expecting an additional short product with sensing stop = 2017-02-25T09:47:03.747.

Can you please check?

**Comment by mivermeu** [comment]
It seems in this case there is a slice boundary before the end of the sensing time, as determined from ANX information. In previous implementations as well as in the current one, framing happens on a single slice, causing the last part of this particular dataset to be cut off. The philosophy behind this is that the input product comprises a slice and should be within slice bounds itself.

We can change this behaviour, if you like, so that frames are constructed from any range of sensing time, no matter where they fall in the slice grid. Unfortunately today is a holiday in the Netherlands and the S&T team will be away till Tuesday. I hope that the current behaviour is not too obstructive to your testing till then.

**Comment by lmar76** [comment]
According to the ANX information:

```json
"anx": [
    "2017-02-25T05:00:18.883013Z",
    "2017-02-25T06:38:29.864514Z",
    "2017-02-25T08:16:40.846015Z",
    "2017-02-25T09:54:51.827516Z",
    "2017-02-25T11:33:02.809016Z",
    "2017-02-25T13:11:13.790517Z"
]
```

The input L0 slice (`S3_RAW__0S`), with sensing interval [2017-02-25T09:46:04.620, 2017-02-25T09:47:03.747] falls in the orbit [2017-02-25T08:16:40.846015Z, 2017-02-25T09:54:51.827516Z] and it is **partial** (i.e. it is the first slice of the data take), because the theoretical slice (number 57) should have sensing interval [2017-02-25T09:45:16.731248440, 2017-02-25 09:47:03.747056180]:

```
start = anx + (57 - 1) * slice_duration - 5
stop = start + slice_duration + 7
```

with `slice_duration = 95.01580774` and `anx = 2017-02-25T08:16:40.846015` but the data take starts at 2017-02-25T09:46:04.620, so the _actual_ slice starts later than the theoretic one but ends at the same time. Why do you say that there is a "slice boundary before the end of the sensing time"?

**Comment by lmar76** [comment]
Hello, I have tried again. Now procsim generates 4 products as expected. The start/stop sensing times seem correct:

| # | Frame ID | File                                                              | Start                      | Stop                       |
|---|:--------:|-------------------------------------------------------------------| ---------------------------|----------------------------|
| 1 | 285      | BIO_OPER_CPF_L1VFRA_20170225T094604_20170225T094620_01_BPF2FO.EOF | 2017-02-25T09:46:04.620000 | 2017-02-25T09:46:20.740861 | 
| 2 | 286      | BIO_OPER_CPF_L1VFRA_20170225T094618_20170225T094639_01_BPF2FO.EOF | 2017-02-25T09:46:18.740861 | 2017-02-25T09:46:39.744023 |
| 3 | 287      | BIO_OPER_CPF_L1VFRA_20170225T094637_20170225T094658_01_BPF2FO.EOF | 2017-02-25T09:46:37.744023 | 2017-02-25T09:46:58.747185 |
| 4 | 288      | BIO_OPER_CPF_L1VFRA_20170225T094656_20170225T094703_01_BPF2FO.EOF | 2017-02-25T09:46:56.747185 | 2017-02-25T09:47:03.747000 |

but I don't have the same values of the Frame ID, i.e. all the `frame_id` in the products seem to be greater then the expected one. As explained above, the orbit interval is [2017-02-25T08:16:40.846015Z, 2017-02-25T09:54:51.827516Z], thus, if you consider frame id 285:

```python
anx = datetime(2017, 2, 25, 8, 16, 40, 846015)
frame_id = 285
frame_duration = 21.00316155  # according to production model
frame_final_overlap = 2
frame_grid_duration = frame_duration - frame_final_overlap
frame_start_time = anx + timedelta(seconds=(frame_id - 1) * frame_grid_duration)
frame_stop_time = anx + timedelta(seconds=frame_id  * frame_grid_duration + frame_final_overlap)
```

this gives:

```python
(datetime.datetime(2017, 2, 25, 9, 46, 37, 743895), datetime.datetime(2017, 2, 25, 9, 46, 58, 747057))
```
which is the frame start/stop values of file #4. It seems that the frame IDs in the products are all "shifted" by two, i.e.: product #1 should have frame id 283, product #2 should have frame id 284, etc.

**Comment by mivermeu** [comment]
You are completely right. In this particular case an error occurred where the frame ID was shifted by two due to a mistake in the calculation of the first frame ID in a slice. This was resolved in PR #40 and merged into the master branch.

**Comment by lmar76** [comment]
According to the discussion at #42, the last generated frame should be even shorter than the one indicated above (see last line):

|   n | start                         | stop                          |   duration |
|----:|:------------------------------|:------------------------------|-----------:|
| 283 | 2017-02-25 09:46:04.620000    | 2017-02-25 09:46:20.740733084 |    16.1207 |
| 284 | 2017-02-25 09:46:18.740733084 | 2017-02-25 09:46:39.743894632 |    21.0032 |
| 285 | 2017-02-25 09:46:37.743894632 | 2017-02-25 09:46:58.747056180 |    21.0032 |
| 286 | 2017-02-25 09:46:56.747056180 | 2017-02-25 09:46:58.747056180 |     2      |

Do you agree?

**Comment by lmar76** [comment]
Hi Milo, in light of https://github.com/stcorp/procsim/issues/42#issuecomment-1152274311, the last product #4 (see list above) should not be generated (because will be generated by processing the nest slice). Please, let me know your opinion.

**Comment by mivermeu** [comment]
Indeed, in your very helpful and thorough description of the situation in https://github.com/stcorp/procsim/issues/42#issuecomment-1152274311, the last part of this particular slice is not transferred to a frame. I agree with this approach in broad lines, although there are subtleties for which I might need your input. In order to make issue tracking a little easier I recommend that we move further discussion on this topic to issue #42.

**Comment by lmar76** [comment]
I think we can close this issue in favor of #42 .

**Comment by lmar76** [comment]
Closed in favor of #42 .
